File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> ch = new HashSet<>();

        int max = 0, l = 0, r = 0;

        while(r < s.length())
        {
            if(!ch.contains(s.charAt(r)))
            {
                ch.add(s.charAt(r));
                max = Math.max(max, ch.size());
                r++;
            }
            else
            {
                ch.remove(s.charAt(l));
                l++;
            }
        }

        return max;
        
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ch = set()
        l = r = max_len = 0

        while r < len(s):
            if s[r] not in ch:
                ch.add(s[r])
                max_len = max(max_len, len(ch))
                r += 1
            else:
                ch.remove(s[l])
                l += 1
        return max_len